core/lib/via_withdrawal_client/src/client.rs

```diff
@@ -0,0 +1,148 @@
+use std::{collections::HashMap, str::FromStr};
+
+use zksync_da_client::DataAvailabilityClient;
+use zksync_types::{web3::keccak256, H160, H256};
+
+use crate::{
+    pubdata::Pubdata,
+    types::{L2BridgeLogMetadata, WithdrawalRequest, L2_BASE_TOKEN_SYSTEM_CONTRACT_ADDR},
+    withdraw::parse_l2_withdrawal_message,
+};
+
+#[derive(Debug)]
+pub struct WithdrawalClient {
+    client: Box<dyn DataAvailabilityClient>,
+}
+
+impl WithdrawalClient {
+    pub fn new(client: Box<dyn DataAvailabilityClient>) -> Self {
+        Self { client }
+    }
+
+    pub async fn get_withdrawals(&self, blob_id: &str) -> anyhow::Result<Vec<WithdrawalRequest>> {
+        let pubdata_bytes = self._fetch_pubdata(blob_id).await?;
+        let pubdata = Pubdata::decode_pubdata(pubdata_bytes)?;
+        let l2_bridge_metadata = WithdrawalClient::_list_l2_bridge_metadata(&pubdata);
+        let withdrawals = WithdrawalClient::_get_valid_withdrawals(l2_bridge_metadata);
+        Ok(withdrawals)
+    }
+
+    async fn _fetch_pubdata(&self, blob_id: &str) -> anyhow::Result<Vec<u8>> {
+        let response = self.client.get_inclusion_data(blob_id).await?;
+        if let Some(inclusion_data) = response {
+            return Ok(inclusion_data.data);
+        };
+        Ok(Vec::new())
+    }
+
+    fn _l2_to_l1_messages_hashmap(pubdata: &Pubdata) -> HashMap<H256, Vec<u8>> {
+        let mut hashes: HashMap<H256, Vec<u8>> = HashMap::new();
+        for message in &pubdata.l2_to_l1_messages {
+            let hash = H256::from(keccak256(&message));
+            hashes.insert(hash, message.clone());
+        }
+        hashes
+    }
+
+    fn _list_l2_bridge_metadata(pubdata: &Pubdata) -> Vec<L2BridgeLogMetadata> {
+        let mut withdrawals: Vec<L2BridgeLogMetadata> = Vec::new();
+        let l2_bridges_hash =
+            H256::from(H160::from_str(L2_BASE_TOKEN_SYSTEM_CONTRACT_ADDR).unwrap());
```

Comment:
don't we need a config for that? system_conntract_addrs will be the same for any chain?

Reply:
never mind, it seems ok